File: testdata/fixtures/typescript/CWE-20.ts
// CWE-20: req.query without validation
import fs from "fs";
import { Request } from "express";

export function vuln1(req: Request) {
  const q = req.query.q as string;
  // ruleid: typescript-20-validation
  parseInt(q);
}

export function vuln2(req: Request) {
  const id = req.params.id as string;
  // ruleid: typescript-20-validation
  Number(id);
}

export function vuln3(req: Request) {
  const p = req.query.path as string;
  // ruleid: typescript-20-validation
  fs.readFileSync(p);
}

export function safeAllowlist(req: Request) {
  const allowlist = req.query.q as string;
  if (!/^[a-z]+$/.test(allowlist)) return;
  // ok: typescript-20-validation
  parseInt(allowlist);
}

export function safeFixed() {
  // ok: typescript-20-validation
  parseInt("42");
}

File: testdata/fixtures/kotlin/CWE-20.kt
// CWE-20: Improper input validation — request query without validation
package fixture

import java.io.File
import javax.servlet.http.HttpServletRequest

class TestCWE20 {
    fun vuln1(request: HttpServletRequest) {
        val q = request.getParameter("q")
        // ruleid: kotlin-20-validation
        File(q)
    }
    fun vuln2(request: HttpServletRequest) {
        val id = request.getParameter("id")
        // ruleid: kotlin-20-validation
        id.toInt()
    }
    fun vuln3(request: HttpServletRequest) {
        val p = request.getParameter("path")
        // ruleid: kotlin-20-validation
        Integer.parseInt(p)
    }
    fun safeAllowlist(request: HttpServletRequest) {
        val allowlist = request.getParameter("q")
        if (!allowlist.matches(Regex("[a-z]+"))) return
        // ok: kotlin-20-validation
        File(allowlist)
    }
    fun safeFixed() {
        // ok: kotlin-20-validation
        File("/tmp/fixed")
    }
}
